Add options flow to the airq integration

Sibgatulin opened the pull request:
Another iteration of a [previous PR](https://github.com/home-assistant/core/pull/102802), which needed to be split.

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Expose to the user the following configuration:
1. A choice between fetching from the device either: a. the averaged (previous and the new default behaviour) or b. noisy momentary sensor reading
2. A toggle to clip (spuriously) negative sensor values (default functionality, previously unexposed)

To those ends:
- Introduce an `OptionsFlowHandler` alongside with a listener `AirQCoordinator.async_set_options`
- Introduce constants to handle represent options
- Add tests and strings

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/31193

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Sibgatulin committed:
Add support for options to airq integration

Expose to the user the following configuration:
1. A choice between fetching from the device either:
    a. the averaged (previous and the new default behaviour) or
    b. noisy momentary sensor reading
2. A toggle to clip (spuriously) negative sensor values (default
    functionality, previously unexposed)

To those ends:
- Introduce an `OptionsFlowHandler` alongside with a listener
  `AirQCoordinator.async_set_options`
- Introduce constants to handle represent options
- Add tests and strings

ludeeus commented on homeassistant/components/airq/config_flow.py:
```diff
@@ -72,3 +74,40 @@ async def async_step_user(
+            return self.async_create_entry(title="", data=self.options)
+        return self.async_show_form(
+            step_id="init",
+            data_schema=vol.Schema(
```
You can utilize the `homeassistant.helpers.schema_config_entry_flow.SchemaOptionsFlowHandler` class for these simple schema flows.
An example of how that can be implemented can be found here <https://github.com/home-assistant/core/blob/dev/homeassistant/components/traccar_server/config_flow.py#L196-L202>

Sibgatulin replied:
Brilliant, thanks!

Sibgatulin commented on homeassistant/components/airq/coordinator.py:
```diff
@@ -56,4 +68,20 @@ async def _async_update_data(self) -> dict:
+            clip_negative_values=self.options[CONF_CLIP_NEGATIVE],
+        )
+
+    async def async_set_options(
+        self, hass: HomeAssistant, config_entry: ConfigEntry
+    ) -> None:
+        """Update the configuration options for the device."""
+        options = self.options | config_entry.options
+        _LOGGER.debug(
+            "%s.async_set_options: from %s to %s",
+            self.__class__.__name__,
+            repr(self.options),
+            repr(options),
+        )
+        self.options = options
```
thanks for spotting, it is indeed a remnant from the previous pattern, you have rightfully discouraged. Removing it.

Sibgatulin commented on homeassistant/components/airq/config_flow.py:
```diff
@@ -72,3 +74,40 @@ async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Manage the options."""
+        if user_input is not None:
```
Looks like I can skip the entire `OptionsFlowHandler` class

Sibgatulin committed:
Drop OptionsFlowHandler in favour of SchemaOptionsFlowHandler

Modify `AirQCoordinator.__init__` to accommodate the change in option
handling, and drop `async_set_options` which slipped through the
previous commit.

emontnemery reviewed (changes requested):
The changes look good, but the PR needs to be rebased.

Sibgatulin committed:
Ruff formatting

Sibgatulin committed:
Merge branch 'dev' into airq_add_option_flow

emontnemery reviewed (approved):
LGTM, thanks @Sibgatulin 👍
